src/v/cloud_topics/dl_stm/dl_stm_state.h

```diff
@@ -0,0 +1,61 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#pragma once
+
+#include "cloud_topics/dl_overlay.h"
+#include "cloud_topics/dl_stm/dl_version.h"
+#include "container/fragmented_vector.h"
+
+namespace experimental::cloud_topics {
+
+struct dl_overlay_entry {
+    dl_overlay overlay;
+
+    dl_version added_at;
+    dl_version removed_at;
+};
+
+/// In-memory state of the data layout state machine (dl_stm).
+///
+/// Separating the state from the state machine allows the state to be
+/// checkpointed and restored independently of the state machine.
+class dl_stm_state {
+    friend class dl_stm_state_accessor;
+
+public:
+    /// Add a new overlay to the state. The overlay becomes visible
+    /// starting with the current version.
+    void push_overlay(dl_overlay overlay);
+
+    /// Find an overlay that contains the given offset. If no overlay
+    /// contains the offset, find the overlay covering the next closest
+    /// available offset.
+    std::optional<dl_overlay> lower_bound(kafka::offset offset) const;
+
+    void set_current_version(dl_version version) {
+        // Greater or equal is required to handle retries.
+        vassert(
+          version >= _current_version,
+          "Version can't go backwards. Current version: {}, new version: {}",
+          _current_version,
+          version);
+
+        _current_version = version;
+    }
+
+private:
+    dl_version _current_version;
+
+    // A list of overlays that are stored in the cloud storage.
+    // The order of elements is undefined.
+    std::deque<dl_overlay_entry> _overlays;
```

Comment:
Goed

Reply:
Please elaborate. Not sure I understand your message.